src/sync/treiber_stack.rs

```diff
@@ -1,82 +1,107 @@
+//! Treiber stacks.
+
 use std::sync::atomic::Ordering::{Acquire, Release, Relaxed};
 use std::ptr;
 
-use mem::epoch::{self, Atomic, Owned};
+use epoch::{self, Atomic};
 
 /// Treiber's lock-free stack.
 ///
-/// Usable with any number of producers and consumers.
+/// This is usable with any number of producers and consumers.
 #[derive(Debug)]
 pub struct TreiberStack<T> {
+    /// The head (top) node.
     head: Atomic<Node<T>>,
 }
 
+/// A node in the stack.
 #[derive(Debug)]
 struct Node<T> {
+    /// The payload of this node.
     data: T,
+    /// The node below this node.
     next: Atomic<Node<T>>,
 }
 
 impl<T> TreiberStack<T> {
     /// Create a new, empty stack.
     pub fn new() -> TreiberStack<T> {
-        TreiberStack { head: Atomic::null() }
+        TreiberStack {
+            head: Atomic::null(),
+        }
     }
 
-    /// Push `t` on top of the stack.
-    pub fn push(&self, t: T) {
-        let mut n = Owned::new(Node {
-            data: t,
+    /// Push an element on top of the stack.
+    pub fn push(&self, elem: T) {
+        // Construct the node.
+        let mut n = Box::new(Node {
+            data: elem,
             next: Atomic::null(),
         });
+
+        // Pin the epoch.
         let guard = epoch::pin();
+        // In order to solve the ABA condition, we spin until the CAS succeeds.
         loop {
+            // Load the head node.
             let head = self.head.load(Relaxed, &guard);
```

Comment:
Again we load the head twice when we already got it from compare and swap. Also we should use pause on x86.